File: tests/workflow/producer/rust/src/producer/responses/structd.rs
use super::{identification, producer::Control, protocol, scope::Scope, Context};
use crate::stat::Alias;
use crate::test::samples;
use clibri::server;

#[allow(unused_variables)]
pub async fn response<'c, E: server::Error, C: server::Control<E>>(
    request: &protocol::StructD,
    scope: &mut Scope<'_, E, C>,
) -> Result<protocol::StructA, protocol::StructC> {
    let index = scope
        .context
        .get_index(scope.identification.uuid(), Alias::StructD);
    if index == 1 {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::StructA);
        Ok(samples::struct_a::get())
    } else {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::StructC);
        Err(samples::struct_c::get())
    }
}

File: tests/workflow/producer/typescript/src/producer/responses/structd.ts
import { Response } from "../implementation/responses/structd";
import {
	Context,
	Producer,
	Identification,
	Filter,
	Protocol,
} from "../implementation/responses";
import { Alias } from "../../stat";
import { panic } from "../../tools";
import { Scope } from "../implementation/scope";

export function response(
	request: Protocol.StructD,
	scope: Scope
): Promise<Response> {
	const stat = scope.context.getStat(scope.consumer.uuid());
	stat.case(Alias.StructD);
	const index = stat.index(Alias.StructD);
	if (index === 0) {
		return Promise.resolve(new Response(Protocol.StructC.defaults()));
	} else if (index === 1) {
		return Promise.resolve(new Response(Protocol.StructA.defaults()));
	} else {
		panic(`Unexpected index for Protocol.StructD request: ${index}`);
		return Promise.reject(
			new Error(`Handler for Protocol.StructD isn't implemented.`)
		);
	}
}
